Tycho doesn't honor 'limit-modules' from '.classpath'

I have a Plug-in Project with the following `.classpath`:

```
<?xml version="1.0" encoding="UTF-8"?>
<classpath>
	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11">
		<attributes>
			<attribute name="module" value="true"/>
			<attribute name="limit-modules" value="java.management,java.xml"/>
		</attributes>
	</classpathentry>
	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
	<classpathentry kind="src" path="src"/>
	<classpathentry kind="output" path="target/classes"/>
</classpath>
```

Note the `limit-modules`, which reduces the JDK modules available, to prevent a JDK module and a plug-in dependency from both providing the same package. This makes sure that I don't have any compile errors in Eclipse when JDT compiles the Java code.

However, for the Maven/Tycho build, it failed again due to a package being provided by both a JDK module and a plug-in dependency.

I had to add the following to the `pom.xml` to resolve this:

```
    <build>
        <plugins>
            <plugin>
                <groupId>org.eclipse.tycho</groupId>
                <artifactId>tycho-compiler-plugin</artifactId>
                <version>${tycho.version}</version>
                <configuration>
                    <compilerArgs>
                        <arg>--add-modules</arg>
                        <arg>java.base,java.management,java.xml</arg>
                        <arg>--limit-modules</arg>
                        <arg>java.base,java.management,java.xml</arg>
                    </compilerArgs>
                </configuration>
            </plugin>
        </plugins>
    </build>
```

Essentially this has the same information as in `.classpath` (`java.base` is omitted there as it is a dependency of one of the other modules).

I'm no expert on how Tycho works under the hood. I think it would be good for Tycho to honor the `limit-modules` information from the `.classpath` file.

[compiler.limit.modules.zip](https://github.com/eclipse/tycho/files/6489807/compiler.limit.modules.zip)

This zip contains a reproducer.

When I execute ` mvn clean verify -Dtycho-version=2.3.0`, using JDK 11, I get:

```
...
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.3.0:compile (default-compile) on project bundle: Compilation failure: Compilation failure:
[ERROR] C:\Users\dhendri1\Desktop\compiler.limit.modules\bundle\src\org\w3c\dom\SomeClass.java:[1]
[ERROR]         package org.w3c.dom;
[ERROR]                 ^^^^^^^^^^^
[ERROR] The package org.w3c.dom conflicts with a package accessible from another module: java.xml
[ERROR] 1 problem (1 error)
...
```

The bundle provides package `org.w3c.dom`, as does JDK module `java.xml`. This is not allowed in Java 11.

By replacing `bundle/pom.xml` by `bundle/pom-to-fix-build.xml`, the build succeeds.

--- a/tycho-compiler-plugin/src/main/java/org/eclipse/tycho/compiler/AbstractOsgiCompilerMojo.java
+++ b/tycho-compiler-plugin/src/main/java/org/eclipse/tycho/compiler/AbstractOsgiCompilerMojo.java
@@ -70,6 +70,8 @@ import org.eclipse.tycho.classpath.SourcepathEntry;
 import org.eclipse.tycho.core.BundleProject;
 import org.eclipse.tycho.core.TychoConstants;
 import org.eclipse.tycho.core.TychoProject;
+import org.eclipse.tycho.core.dotClasspath.JREClasspathEntry;
+import org.eclipse.tycho.core.dotClasspath.ProjectClasspathEntry;
 import org.eclipse.tycho.core.ee.ExecutionEnvironmentUtils;
 import org.eclipse.tycho.core.ee.StandardExecutionEnvironment;
 import org.eclipse.tycho.core.ee.shared.ExecutionEnvironment;
@@ -567,6 +569,18 @@ public abstract class AbstractOsgiCompilerMojo extends AbstractCompilerMojo
         configureJavaHome(compilerConfiguration);
         configureBootclasspathAccessRules(compilerConfiguration);
         configureCompilerLog(compilerConfiguration);
+        Collection<ProjectClasspathEntry> classpathEntries = getEclipsePluginProject().getClasspathEntries();
+        for (ProjectClasspathEntry cpe : classpathEntries) {
+            if (cpe instanceof JREClasspathEntry) {
+                JREClasspathEntry jreClasspathEntry = (JREClasspathEntry) cpe;
+                if (jreClasspathEntry.isModule()) {
+                    Collection<String> modules = jreClasspathEntry.getLimitModules();
+                    if (modules.size() > 0) {
+                        compilerConfiguration.addCompilerCustomArgument("--limit-modules", String.join(",", modules));
+                    }
+                }
+            }
+        }
         return compilerConfiguration;
     }
 

--- a/tycho-its/projects/compiler.limit.modules/bundle/.classpath
+++ b/tycho-its/projects/compiler.limit.modules/bundle/.classpath
@@ -1,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-11">
+	<attributes>
+		<attribute name="module" value="true"/>
+		<attribute name="limit-modules" value="java.base"/>
+	</attributes>
+</classpathentry>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src"/>
+	<classpathentry kind="output" path="target/classes"/>
+</classpath>

--- a/tycho-its/projects/compiler.limit.modules/bundle/META-INF/MANIFEST.MF
+++ b/tycho-its/projects/compiler.limit.modules/bundle/META-INF/MANIFEST.MF
@@ -1,0 +1,8 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Bundle Plug-in
+Bundle-SymbolicName: bundle.jre.modules
+Bundle-Version: 0.0.1.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-11
+Export-Package: org.w3c.dom
+Automatic-Module-Name: bundle

--- a/tycho-its/projects/compiler.limit.modules/bundle/build.properties
+++ b/tycho-its/projects/compiler.limit.modules/bundle/build.properties
@@ -1,0 +1,4 @@
+source.. = src/
+output.. = target/classes
+bin.includes = META-INF/,\
+               .

--- a/tycho-its/projects/compiler.limit.modules/bundle/pom.xml
+++ b/tycho-its/projects/compiler.limit.modules/bundle/pom.xml
@@ -1,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.eclipse.tycho.tycho-its.compiler.limit.modules</groupId>
+        <artifactId>parent</artifactId>
+        <version>0.0.1-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>bundle.jre.modules</artifactId>
+    <packaging>eclipse-plugin</packaging>
+</project>

--- a/tycho-its/projects/compiler.limit.modules/bundle/src/org/w3c/dom/SomeClass.java
+++ b/tycho-its/projects/compiler.limit.modules/bundle/src/org/w3c/dom/SomeClass.java
@@ -1,0 +1,20 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Dennis Hendriks and others.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Dennis Hendriks - initial API and implementation
+ *******************************************************************************/
+
+package org.w3c.dom;
+
+// Some class in org.w3c.dom package.
+// JDK module java.xml module also provides this package.
+// Java 11 requires 'limit-modules' to exclude java.xml module from JDK.
+public class SomeClass {
+}

--- a/tycho-its/projects/compiler.limit.modules/pom.xml
+++ b/tycho-its/projects/compiler.limit.modules/pom.xml
@@ -1,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.eclipse.tycho.tycho-its.compiler.limit.modules</groupId>
+    <artifactId>parent</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+    <packaging>pom</packaging>
+
+    <modules>
+        <module>bundle</module>
+    </modules>
+
+    <build>
+		<plugins>
+			<plugin>
+				<groupId>org.eclipse.tycho</groupId>
+				<artifactId>tycho-maven-plugin</artifactId>
+				<version>${tycho-version}</version>
+				<extensions>true</extensions>
+			</plugin>
+		</plugins>
+	</build>
+</project>

--- a/tycho-its/src/test/java/org/eclipse/tycho/test/compiler/MavenCompilerPluginTest.java
+++ b/tycho-its/src/test/java/org/eclipse/tycho/test/compiler/MavenCompilerPluginTest.java
@@ -40,4 +40,11 @@ public class MavenCompilerPluginTest extends AbstractTychoIntegrationTest {
         verifier.verifyErrorFreeLog();
     }
 
+    @Test
+    public void testLimitModules() throws Exception {
+        Verifier verifier = getVerifier("compiler.limit.modules", true);
+        verifier.executeGoal("compile");
+        verifier.verifyErrorFreeLog();
+    }
+
 }

